midend/interpreter.cpp

```diff
@@ -878,6 +878,25 @@ void ExpressionEvaluator::postorder(const IR::Operation_Unary* expression) {
         DoConstantFolding cf(refMap, typeMap);
         cf.setCalledBy(this);
         auto result = expression->apply(cf);
+        if (!result->is<IR::BoolLiteral>()) {
```

Comment:
are you positive that this case can happen?
I would think that this would be impossible - the result of constant folding a boolean expression has to a boolean. You are folding here expression, not l.
Which one of the tests exercises this case?

Reply:
You have the same bug here: the previous statement should be `clone->apply(cf)`.
Then all this code is no longer needed.